services/QuillLMS/app/services/independent_practice_activity_packs_assigner.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+class IndependentPracticeActivityPacksAssigner < ApplicationService
+  attr_reader :assigning_all_recommendations,
+    :classroom_id,
+    :diagnostic_activity_id,
+    :release_method,
+    :selections,
+    :user
+
+  class TeacherNotAssociatedWithClassroomError < StandardError; end
+
+  def initialize(
+    assigning_all_recommendations:,
+    classroom_id:,
+    diagnostic_activity_id:,
+    release_method:,
+    selections:,
+    user:
+  )
+    @assigning_all_recommendations = assigning_all_recommendations
+    @classroom_id = classroom_id
+    @diagnostic_activity_id = diagnostic_activity_id
+    @release_method = release_method
+    @selections = selections
+    @user = user
+  end
+
+  def run
+    raise TeacherNotAssociatedWithClassroomError unless teaches_classroom?
+
+    find_or_create_activity_packs
+
+    return if selections_with_students.empty?
+
+    set_diagnostic_recommendations_start_time
+    assign_recommendations
+  end
+
+  private def activity_pack_sequence
+    @activity_pack_sequence ||= ActivityPackSequenceGetter.run(classroom_id, diagnostic_activity_id, release_method)
+  end
+
+  private def assign_recommendations
+    selections_with_students.each_with_index do |selection, index|
+      AssignRecommendationsWorker.perform_async(
+        activity_pack_sequence_id: activity_pack_sequence&.id,
+        assigning_all_recommendations: assigning_all_recommendations,
+        classroom_id: classroom_id,
+        is_last_recommendation: index == last_recommendation_index,
+        lesson: false,
+        order: selection[:order],
+        student_ids: selection[:classrooms][0][:student_ids].compact,
+        unit_template_id: selection[:id]
+      )
+    end
+  end
+
+  private def find_or_create_activity_packs
+    selections
+      .map { |selection| UnitTemplate.find(selection['id']) }
+      .each { |ut| Unit.unscoped.find_or_create_by!(unit_template: ut, name: ut.name, user: user) }
```

Comment:
Bang method in a loop. What do we want to happen here if the 3rd of 20 `units` fails. If they all should fail, we likely need a transaction somewhere.

Reply:
Good call!